Platform compatibility analyzer does not work for property initializer syntax

```C#
class Program
{
    public static object Property { get; } = OperatingSystem.IsWindows() ? null : new Foo();
}

[UnsupportedOSPlatform("windows")]
class Foo
{
}
```

This fails with

```
Program.cs(9,87,9,96): warning CA1416: 'Foo' is unsupported on 'windows'
```

SDK version: 5.0.100-rc.1.20431.3

--- a/src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/InteropServices/PlatformCompatabilityAnalyzer.cs
+++ b/src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/InteropServices/PlatformCompatabilityAnalyzer.cs
@@ -173,13 +173,13 @@ namespace Microsoft.NetCore.Analyzers.InteropServices
                         return;
                     }
 
-                    if (guardMethods.IsEmpty || !(context.OperationBlocks.GetControlFlowGraph(out var topmostBlock) is { } cfg))
+                    if (guardMethods.IsEmpty || !(context.OperationBlocks.GetControlFlowGraph() is { } cfg))
                     {
                         ReportDiagnosticsForAll(platformSpecificOperations, context);
                         return;
                     }
 
-                    var performValueContentAnalysis = ComputeNeedsValueContentAnalysis(topmostBlock!, guardMethods);
+                    var performValueContentAnalysis = ComputeNeedsValueContentAnalysis(cfg.OriginalOperation, guardMethods);
                     var wellKnownTypeProvider = WellKnownTypeProvider.GetOrCreate(context.Compilation);
                     var analysisResult = GlobalFlowStateAnalysis.TryGetOrComputeResult(
                         cfg, context.OwningSymbol, CreateOperationVisitor, wellKnownTypeProvider,
@@ -242,7 +242,7 @@ namespace Microsoft.NetCore.Analyzers.InteropServices
             return false;
         }
 
-        private static bool ComputeNeedsValueContentAnalysis(IBlockOperation operationBlock, ImmutableArray<IMethodSymbol> guardMethods)
+        private static bool ComputeNeedsValueContentAnalysis(IOperation operationBlock, ImmutableArray<IMethodSymbol> guardMethods)
         {
             foreach (var operation in operationBlock.Descendants())
             {

--- a/src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Security/UseContainerLevelAccessPolicy.cs
+++ b/src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Security/UseContainerLevelAccessPolicy.cs
@@ -137,8 +137,7 @@ namespace Microsoft.NetCore.Analyzers.Security
 
                                 if (argumentOperation != null)
                                 {
-                                    var cfg = invocationOperation.GetTopmostParentBlock()?.GetEnclosingControlFlowGraph();
-                                    if (cfg != null)
+                                    if (invocationOperation.TryGetEnclosingControlFlowGraph(out var cfg))
                                     {
                                         var interproceduralAnalysisConfig = InterproceduralAnalysisConfiguration.Create(
                                                                                 operationAnalysisContext.Options,

--- a/src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Security/UseSharedAccessProtocolHttpsOnly.cs
+++ b/src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Security/UseSharedAccessProtocolHttpsOnly.cs
@@ -135,8 +135,7 @@ namespace Microsoft.NetCore.Analyzers.Security
 
                             if (protocolsArgumentOperation != null)
                             {
-                                var cfg = invocationOperation.GetTopmostParentBlock()?.GetEnclosingControlFlowGraph();
-                                if (cfg != null)
+                                if (invocationOperation.TryGetEnclosingControlFlowGraph(out var cfg))
                                 {
                                     var interproceduralAnalysisConfig = InterproceduralAnalysisConfiguration.Create(
                                                                         operationAnalysisContext.Options,

--- a/src/NetAnalyzers/UnitTests/Microsoft.NetCore.Analyzers/InteropServices/PlatformCompatabilityAnalyzerTests.GuardedCallsTests.cs
+++ b/src/NetAnalyzers/UnitTests/Microsoft.NetCore.Analyzers/InteropServices/PlatformCompatabilityAnalyzerTests.GuardedCallsTests.cs
@@ -2,6 +2,7 @@
 
 using System.Collections.Generic;
 using System.Threading.Tasks;
+using Test.Utilities;
 using Xunit;
 
 using VerifyCS = Test.Utilities.CSharpCodeFixVerifier<
@@ -1130,6 +1131,86 @@ public class Test
     {
         return option;
     }
+}
+" + MockAttributesCsSource + MockOperatingSystemApiSource;
+            await VerifyAnalyzerAsyncCs(source, s_msBuildPlatforms);
+        }
+
+        [Fact, WorkItem(4105, "https://github.com/dotnet/roslyn-analyzers/issues/4105")]
+        public async Task OsDependentPropertyWithInitializer_NoDiagnostic()
+        {
+            var source = @"
+using System.Runtime.Versioning;
+using System;
+
+class C
+{
+    public static object Property { get; } = OperatingSystemHelper.IsWindows() ? null : new A();
+}
+
+[UnsupportedOSPlatform(""windows"")]
+class A
+{
+}
+" + MockAttributesCsSource + MockOperatingSystemApiSource;
+            await VerifyAnalyzerAsyncCs(source, s_msBuildPlatforms);
+        }
+
+        [Fact, WorkItem(4105, "https://github.com/dotnet/roslyn-analyzers/issues/4105")]
+        public async Task OsDependentPropertyWithInitializer_Diagnostic()
+        {
+            var source = @"
+using System.Runtime.Versioning;
+using System;
+
+class C
+{
+    public static object Property { get; } = OperatingSystemHelper.IsWindows() ? [|new A()|] : null;
+}
+
+[UnsupportedOSPlatform(""windows"")]
+class A
+{
+}
+" + MockAttributesCsSource + MockOperatingSystemApiSource;
+            await VerifyAnalyzerAsyncCs(source, s_msBuildPlatforms);
+        }
+
+        [Fact, WorkItem(4105, "https://github.com/dotnet/roslyn-analyzers/issues/4105")]
+        public async Task OsDependentFieldWithInitializer_NoDiagnostic()
+        {
+            var source = @"
+using System.Runtime.Versioning;
+using System;
+
+class C
+{
+    private static object _field = OperatingSystemHelper.IsWindows() ? null : new A();
+}
+
+[UnsupportedOSPlatform(""windows"")]
+class A
+{
+}
+" + MockAttributesCsSource + MockOperatingSystemApiSource;
+            await VerifyAnalyzerAsyncCs(source, s_msBuildPlatforms);
+        }
+
+        [Fact, WorkItem(4105, "https://github.com/dotnet/roslyn-analyzers/issues/4105")]
+        public async Task OsDependentFieldWithInitializer_Diagnostic()
+        {
+            var source = @"
+using System.Runtime.Versioning;
+using System;
+
+class C
+{
+    private static object _field = OperatingSystemHelper.IsWindows() ? [|new A()|] : null;
+}
+
+[UnsupportedOSPlatform(""windows"")]
+class A
+{
 }
 " + MockAttributesCsSource + MockOperatingSystemApiSource;
             await VerifyAnalyzerAsyncCs(source, s_msBuildPlatforms);

--- a/src/Utilities/Compiler/Extensions/IOperationExtensions.cs
+++ b/src/Utilities/Compiler/Extensions/IOperationExtensions.cs
@@ -430,8 +430,9 @@ namespace Analyzer.Utilities.Extensions
                 case IPropertyInitializerOperation propertyInitializerOperation:
                     return ControlFlowGraph.Create(propertyInitializerOperation);
 
-                case IParameterInitializerOperation parameterInitializerOperation:
-                    return ControlFlowGraph.Create(parameterInitializerOperation);
+                case IParameterInitializerOperation:
+                    // We do not support flow analysis for parameter initializers
+                    return null;
 
                 default:
                     // Attribute blocks have OperationKind.None, but ControlFlowGraph.Create does not

--- a/src/Utilities/FlowAnalysis/Extensions/OperationBlocksExtensions.cs
+++ b/src/Utilities/FlowAnalysis/Extensions/OperationBlocksExtensions.cs
@@ -3,7 +3,6 @@
 using System.Collections.Immutable;
 using Microsoft.CodeAnalysis;
 using Microsoft.CodeAnalysis.FlowAnalysis;
-using Microsoft.CodeAnalysis.Operations;
 
 namespace Analyzer.Utilities.Extensions
 {

--- a/src/Utilities/FlowAnalysis/Extensions/OperationBlocksExtensions.cs
+++ b/src/Utilities/FlowAnalysis/Extensions/OperationBlocksExtensions.cs
@@ -9,22 +9,17 @@ namespace Analyzer.Utilities.Extensions
 {
     internal static partial class OperationBlocksExtensions
     {
-        public static ControlFlowGraph? GetControlFlowGraph(this ImmutableArray<IOperation> operationBlocks, out IBlockOperation? topmostBlock)
+        public static ControlFlowGraph? GetControlFlowGraph(this ImmutableArray<IOperation> operationBlocks)
         {
             foreach (var operationRoot in operationBlocks)
             {
-                topmostBlock = operationRoot.GetTopmostParentBlock();
-                if (topmostBlock != null)
+                if (operationRoot.TryGetEnclosingControlFlowGraph(out var cfg))
                 {
-                    return topmostBlock.GetEnclosingControlFlowGraph();
+                    return cfg;
                 }
             }
 
-            topmostBlock = null;
             return null;
         }
-
-        public static ControlFlowGraph? GetControlFlowGraph(this ImmutableArray<IOperation> operationBlocks)
-            => operationBlocks.GetControlFlowGraph(out _);
     }
 }

